ompi/mca/coll/libpnbc_osc/pnbc_osc.c

```diff
@@ -604,6 +644,63 @@ static inline int NBC_Start_round(NBC_Handle *handle) {
 #ifdef NBC_TIMING
         Iget_time += MPI_Wtime();
 #endif
+        break;
+     case TRY_GET:
+        NBC_DEBUG(5,"  TRY_GET (offset %li) ", offset);
+        NBC_GET_BYTES(ptr,trygetargs);
+        NBC_DEBUG(5,"*buf: %p, origin count: %i, origin type: %p, target: %i, target count: %i, target type: %p, tag: %i)\n", trygetargs.buf, trygetargs.origin_count, trygetargs.origin_datatype, trygetargs.target, trygetargs.target_count, trygetargs.target_datatype, handle->tag);
+        /* get an additional request */
+        handle->req_count++;
+        /* get buffer */
+        if(trygetargs.tmpbuf) {
+          buf1=(char*)handle->tmpbuf+(long)trygetargs.buf;
+        } else {
+          buf1=(void *)trygetargs.buf;
+        }
+#ifdef NBC_TIMING
+        Iget_time -= MPI_Wtime();
+#endif
+        //TODO: I am not too sure we need to realloc for PUT/GET - Not used
+        tmp = (MPI_Request *) realloc ((void *) handle->req_array, 
+                                       handle->req_count * sizeof (MPI_Request));
+        if (NULL == tmp) {
+          return OMPI_ERR_OUT_OF_RESOURCE;
+        }
+        
+        handle->req_array = tmp;
+        
+        /* [state is unlocked] */
+        res = handle->win->w_osc_module->osc_try_lock(trygetargs.lock_type, trygetargs.target, 
+                                              trygetargs.assert, handle->win);
+        if(OMPI_SUCCESS == res){
```

Comment:
This doesn't fully capture the operation of the state machine.

If the `osc_try_lock` succeeds but then the `osc_get` fails, then the window is still locked. That is a real error situation, so it is unfortunate that the window is now broken but it doesn't matter because MPI will return a real error to the user (and/or abort all processes).

However, if the `osc_try_lock` succeeds and the `osc_get` succeeds, but the `osc_try_unlock` returns `ompi_err_rma_nb_pending`, then the window is already locked and the data is already on its way - the next time around this function should just try the unlock again, without re-doing the lock and get steps.

The state needs to be stored somewhere and retrieved when this executes again. This needs more thought - lets add this to the agenda for a future code design meeting :)